File: src/parsers/javascript_parser.py
import re
import os
from typing import Dict, List, Any, Optional
import logging
from .generic_parser import GenericParser

logger = logging.getLogger(__name__)

class JavaScriptParser(GenericParser):
    """
    Parser for JavaScript files using regex-based parsing.
    For a production implementation, tree-sitter would be more robust.
    """
    
    def __init__(self, file_path: str):
        """
        Initialize the JavaScript parser.
        
        Args:
            file_path: Path to the JavaScript file
        """
        super().__init__(file_path)
        
        # Common patterns
        self.import_pattern = re.compile(
            r'(?:import\s+{([^}]*)}\s+from\s+[\'"]([^\'"]*)[\'"])|'  # import { something } from 'module'
            r'(?:import\s+(\w+)\s+from\s+[\'"]([^\'"]*)[\'"])|'      # import something from 'module'
            r'(?:import\s+\*\s+as\s+(\w+)\s+from\s+[\'"]([^\'"]*)[\'"])|'  # import * as something from 'module'
            r'(?:import\s+[\'"]([^\'"]*)[\'"])'                      # import 'module'
        )
        
        self.export_pattern = re.compile(
            r'export\s+(?:default\s+)?(?:const|let|var|function|class)\s+(\w+)'
        )
        
        self.function_pattern = re.compile(
            r'(?:function\s+(\w+)\s*\(([^)]*)\))|'                    # function name(params)
            r'(?:(?:const|let|var)\s+(\w+)\s*=\s*(?:async\s*)?\(([^)]*)\)\s*=>)|'  # const name = (params) =>
            r'(?:(?:const|let|var)\s+(\w+)\s*=\s*(?:async\s*)?function\s*\(([^)]*)\))'  # const name = function(params)
        )
        
        self.class_pattern = re.compile(
            r'class\s+(\w+)(?:\s+extends\s+(\w+))?\s*{'
        )
        
        self.method_pattern = re.compile(
            r'(?:async\s+)?(\w+)\s*\(([^)]*)\)\s*{'    # method(params) {
        )
        
        self.react_component_pattern = re.compile(
            r'(?:function\s+(\w+)\s*\((?:props|{[^}]*})\))|'  # function Component(props)
            r'(?:const\s+(\w+)\s*=\s*\((?:props|{[^}]*})\)\s*=>)'  # const Component = (props) =>
        )
        
        self.api_route_pattern = re.compile(
            r'(?:app|router)[.](?:get|post|put|delete|patch)\s*\(\s*[\'"]([^\'"]*)[\'"]'  # Express-style routes
        )
    
    def _extract_jsdoc_comment(self, line_number: int) -> Optional[str]:
        """
        Extract JSDoc comment above a given line.
        
        Args:
            line_number: Line number to look above
            
        Returns:
            JSDoc comment text or None if not found
        """
        if not self.content:
            return None
            
        lines = self.content.splitlines()
        if line_number <= 0 or line_number > len(lines):
            return None
            
        # Look for JSDoc-style comment above the line
        comment_lines = []
        current_line = line_number - 2  # -2 because line_number is 1-indexed, and we want to start one line above
        
        # Find the start of the comment block
        while current_line >= 0 and current_line < len(lines):
            line = lines[current_line].strip()
            if line.endswith('*/'):
                comment_lines.insert(0, line)
                break
            current_line -= 1
            
        # If we didn't find the end marker, no JSDoc comment
        if not comment_lines:
            return None
            
        # Collect the rest of the comment block
        current_line -= 1
        while current_line >= 0 and current_line < len(lines):
            line = lines[current_line].strip()
            comment_lines.insert(0, line)
            if line.startswith('/**'):
                # Clean up the comment
                comment = '\n'.join(comment_lines)
                # Remove comment markers
                comment = re.sub(r'/\*\*|\*/|^\s*\*', '', comment, flags=re.MULTILINE)
                return comment.strip()
            current_line -= 1
            
        return None
    
    def parse(self) -> Dict[str, Any]:
        """
        Parse the JavaScript file and extract detailed information.
        
        Returns:
            Dictionary with parsed information
        """
        if not self.content:
            return {'error': 'Could not read file'}
        
        # Get file summary
        summary = self.extract_file_summary()
        
        # Get imports
        imports = self.extract_imports()
        
        # Get functions
        functions = self.extract_functions()
        
        # Get classes
        classes = self.extract_classes()
        
        # Extract React components
        react_components = self.extract_react_components()
        
        # Extract API endpoints
        api_endpoints = self.extract_api_endpoints()
        
        return {
            'summary': summary,
            'imports': imports,
            'functions': functions,
            'classes': classes,
            'react_components': react_components,
            'api_endpoints': api_endpoints,
        }
    
    def extract_imports(self) -> List[Dict[str, Any]]:
        """
        Extract import statements from the JavaScript file.
        
        Returns:
            List of import statements with metadata
        """
        if not self.content:
            return []
            
        imports = []
        
        for line_number, line in enumerate(self.content.splitlines(), 1):
            for match in self.import_pattern.finditer(line):
                groups = match.groups()
                
                if groups[0]:  # import { something } from 'module'
                    module = groups[1]
                    items = [item.strip() for item in groups[0].split(',')]
                    for item in items:
                        # Handle "as" aliases
                        if ' as ' in item:
                            original, alias = item.split(' as ')
                            imports.append({
                                'type': 'named_import',
                                'name': original.strip(),
                                'alias': alias.strip(),
                                'module': module,
                                'line_number': line_number,
                            })
                        else:
                            imports.append({
                                'type': 'named_import',
                                'name': item,
                                'module': module,
                                'line_number': line_number,
                            })
                            
                elif groups[2]:  # import something from 'module'
                    imports.append({
                        'type': 'default_import',
                        'name': groups[2],
                        'module': groups[3],
                        'line_number': line_number,
                    })
                    
                elif groups[4]:  # import * as something from 'module'
                    imports.append({
                        'type': 'namespace_import',
                        'name': groups[4],
                        'module': groups[5],
                        'line_number': line_number,
                    })
                    
                elif groups[6]:  # import 'module'
                    imports.append({
                        'type': 'side_effect_import',
                        'module': groups[6],
                        'line_number': line_number,
                    })
        
        return imports
    
    def extract_functions(self) -> List[Dict[str, Any]]:
        """
        Extract function definitions from the JavaScript file.
        
        Returns:
            List of function definitions with metadata
        """
        if not self.content:
            return []
            
        functions = []
        
        for line_number, line in enumerate(self.content.splitlines(), 1):
            for match in self.function_pattern.finditer(line):
                groups = match.groups()
                
                function_name = groups[0] or groups[2] or groups[4]
                params_str = groups[1] or groups[3] or groups[5] or ''
                
                if function_name:
                    params = [param.strip() for param in params_str.split(',') if param.strip()]
                    
                    # Extract JSDoc
                    jsdoc = self._extract_jsdoc_comment(line_number)
                    
                    functions.append({
                        'name': function_name,
                        'params': params,
                        'line_number': line_number,
                        'jsdoc': jsdoc,
                    })
        
        return functions
    
    def extract_classes(self) -> List[Dict[str, Any]]:
        """
        Extract class definitions from the JavaScript file.
        
        Returns:
            List of class definitions with metadata
        """
        if not self.content:
            return []
            
        classes = []
        current_class = None
        
        lines = self.content.splitlines()
        for line_number, line in enumerate(lines, 1):
            # Look for class definitions
            class_match = self.class_pattern.search(line)
            if class_match:
                class_name, parent_class = class_match.groups()
                
                # Extract JSDoc
                jsdoc = self._extract_jsdoc_comment(line_number)
                
                current_class = {
                    'name': class_name,
                    'extends': parent_class,
                    'line_number': line_number,
                    'jsdoc': jsdoc,
                    'methods': [],
                }
                classes.append(current_class)
                continue
                
            # If inside a class, look for methods
            if current_class and '{' in line:
                method_match = self.method_pattern.search(line)
                if method_match:
                    method_name, params_str = method_match.groups()
                    params = [param.strip() for param in params_str.split(',') if param.strip()]
                    
                    # Skip constructor for inherited methods check
                    is_constructor = method_name == 'constructor'
                    
                    # Extract JSDoc
                    jsdoc = self._extract_jsdoc_comment(line_number)
                    
                    current_class['methods'].append({
                        'name': method_name,
                        'params': params,
                        'is_constructor': is_constructor,
                        'line_number': line_number,
                        'jsdoc': jsdoc,
                    })
        
        return classes
    
    def extract_react_components(self) -> List[Dict[str, Any]]:
        """
        Extract React component definitions from the JavaScript file.
        
        Returns:
            List of React component definitions with metadata
        """
        if not self.content:
            return []
            
        components = []
        
        # Check if this file uses React
        if 'import React' not in self.content and 'from "react"' not in self.content and 'from \'react\'' not in self.content:
            return []
            
        for line_number, line in enumerate(self.content.splitlines(), 1):
            match = self.react_component_pattern.search(line)
            if match:
                groups = match.groups()
                component_name = groups[0] or groups[1]
                
                if component_name:
                    # Extract JSDoc
                    jsdoc = self._extract_jsdoc_comment(line_number)
                    
                    # Look for prop types
                    prop_types = []
                    if component_name and f"{component_name}.propTypes" in self.content:
                        # Very basic prop types extraction - a more robust solution would use AST parsing
                        prop_types_pattern = re.compile(f"{component_name}.propTypes\\s*=\\s*{{([^}}]*)}}");
                        prop_types_match = prop_types_pattern.search(self.content)
                        if prop_types_match:
                            prop_types_str = prop_types_match.group(1)
                            # Extract prop names
                            for prop in re.finditer(r'(\w+)\s*:', prop_types_str):
                                prop_types.append(prop.group(1))
                    
                    components.append({
                        'name': component_name,
                        'line_number': line_number,
                        'jsdoc': jsdoc,
                        'prop_types': prop_types,
                    })
        
        return components
    
    def extract_api_endpoints(self) -> List[Dict[str, Any]]:
        """
        Extract API endpoints from the JavaScript file.
        
        Returns:
            List of API endpoints with metadata
        """
        if not self.content:
            return []
            
        endpoints = []
        
        for line_number, line in enumerate(self.content.splitlines(), 1):
            for match in self.api_route_pattern.finditer(line):
                route = match.group(1)
                
                # Extract HTTP method
                method_match = re.search(r'(?:app|router)[.](\w+)\s*\(', line)
                http_method = method_match.group(1).upper() if method_match else 'GET'
                
                endpoints.append({
                    'route': route,
                    'method': http_method,
                    'line_number': line_number,
                })
        
        return endpoints

File: src/parsers/javascript-parser.ts
import { BaseParser } from './base-parser';
import { ParseResult, FunctionInfo, ClassInfo, ImportInfo, ExportInfo, ConstantInfo, ParameterInfo, PropertyInfo, ImportItem } from '../types';

export class JavaScriptParser extends BaseParser {
  public parse(): ParseResult {
    return {
      functions: this.parseFunctions(),
      classes: this.parseClasses(),
      imports: this.parseImports(),
      exports: this.parseExports(),
      constants: this.parseConstants()
    };
  }

  private parseFunctions(): FunctionInfo[] {
    const functions: FunctionInfo[] = [];
    
    // Patterns for different function declarations
    const patterns = [
      /^\s*(?:export\s+)?(?:async\s+)?function\s+(\w+)\s*\(([^)]*)\)\s*\{/,
      /^\s*(?:export\s+)?(?:const|let|var)\s+(\w+)\s*=\s*(?:async\s+)?\(([^)]*)\)\s*=>/,
      /^\s*(?:export\s+)?(?:const|let|var)\s+(\w+)\s*=\s*(?:async\s+)?function\s*\(([^)]*)\)\s*\{/,
      /^\s*(\w+)\s*:\s*(?:async\s+)?\(([^)]*)\)\s*=>/,
      /^\s*(\w+)\s*:\s*(?:async\s+)?function\s*\(([^)]*)\)\s*\{/
    ];

    for (let i = 0; i < this.lines.length; i++) {
      const line = this.lines[i];
      
      for (const pattern of patterns) {
        const match = line.match(pattern);
        if (match) {
          const name = match[1];
          const params = this.convertToParameterInfo(this.parseParameters(match[2] || ''));
          const isAsync = line.includes('async');
          const lineStart = i + 1;
          const lineEnd = this.findFunctionEnd(i);
          const docstring = this.extractDocstring(Math.max(0, i - 3), i);

          functions.push({
            name,
            parameters: params,
            isAsync,
            lineStart,
            lineEnd,
            docstring
          });
          break;
        }
      }
    }

    return functions;
  }

  private parseClasses(): ClassInfo[] {
    const classes: ClassInfo[] = [];
    
    const classPattern = /^\s*(?:export\s+)?class\s+(\w+)(?:\s+extends\s+(\w+))?\s*\{/;

    for (let i = 0; i < this.lines.length; i++) {
      const line = this.lines[i];
      const match = line.match(classPattern);
      
      if (match) {
        const name = match[1];
        const extendsClass = match[2];
        const lineStart = i + 1;
        const lineEnd = this.findEndLine(lineStart, '{', '}');
        const docstring = this.extractDocstring(Math.max(0, i - 3), i);

        const methods = this.parseClassMethods(lineStart, lineEnd);
        const properties = this.convertToPropertyInfo(this.parseClassProperties(lineStart, lineEnd));

        classes.push({
          name,
          methods,
          properties,
          extends: extendsClass,
          lineStart,
          lineEnd,
          docstring
        });
      }
    }

    return classes;
  }

  private parseImports(): ImportInfo[] {
    const imports: ImportInfo[] = [];
    
    const patterns = [
      /^import\s+(\w+)\s+from\s+['"]([^'"]+)['"]/,  // import default from 'module'
      /^import\s+\{\s*([^}]+)\s*\}\s+from\s+['"]([^'"]+)['"]/,  // import { named } from 'module'
      /^import\s+(\w+),\s*\{\s*([^}]+)\s*\}\s+from\s+['"]([^'"]+)['"]/,  // import default, { named } from 'module'
      /^import\s+\*\s+as\s+(\w+)\s+from\s+['"]([^'"]+)['"]/,  // import * as name from 'module'
      /^const\s+(?:\{\s*([^}]+)\s*\}|(\w+))\s*=\s*require\(['"]([^'"]+)['"]\)/  // require syntax
    ];

    for (const line of this.lines) {
      for (const pattern of patterns) {
        const match = line.match(pattern);
        if (match) {
          if (pattern.source.includes('require')) {
            // Handle require syntax
            const items = match[1] ? this.convertToImportItems(match[1].split(',').map(s => s.trim())) : [{name: match[2] || ''}];
            const module = match[3];
            imports.push({
              module,
              items,
              isDefault: !match[1]
            });
          } else {
            // Handle import syntax
            let items: ImportItem[] = [];
            let isDefault = false;
            let module = match[2] || match[3];

            if (match[0].includes('* as')) {
              items = [{name: match[1]}];
            } else if (match[0].includes('{')) {
              const itemsString = match[1] || match[2];
              items = this.convertToImportItems(itemsString.split(',').map(s => s.trim()));
            } else {
              items = [{name: match[1]}];
              isDefault = true;
            }

            imports.push({
              module,
              items,
              isDefault
            });
          }
          break;
        }
      }
    }

    return imports;
  }

  private parseExports(): ExportInfo[] {
    const exports: ExportInfo[] = [];
    
    for (const line of this.lines) {
      // Export function
      if (line.match(/^\s*export\s+(?:async\s+)?function\s+(\w+)/)) {
        const match = line.match(/^\s*export\s+(?:async\s+)?function\s+(\w+)/);
        if (match) {
          exports.push({ name: match[1], type: 'function' });
        }
      }
      
      // Export class
      else if (line.match(/^\s*export\s+class\s+(\w+)/)) {
        const match = line.match(/^\s*export\s+class\s+(\w+)/);
        if (match) {
          exports.push({ name: match[1], type: 'class' });
        }
      }
      
      // Export const/let/var
      else if (line.match(/^\s*export\s+(?:const|let|var)\s+(\w+)/)) {
        const match = line.match(/^\s*export\s+(?:const|let|var)\s+(\w+)/);
        if (match) {
          exports.push({ name: match[1], type: 'constant' });
        }
      }
      
      // Export default
      else if (line.match(/^\s*export\s+default\s+/)) {
        const match = line.match(/^\s*export\s+default\s+(?:class\s+)?(\w+)/);
        exports.push({ 
          name: match ? match[1] : 'default', 
          type: 'default' 
        });
      }
      
      // Export { ... }
      else if (line.match(/^\s*export\s+\{([^}]+)\}/)) {
        const match = line.match(/^\s*export\s+\{([^}]+)\}/);
        if (match) {
          const items = match[1].split(',').map(s => s.trim());
          for (const item of items) {
            exports.push({ name: item, type: 'constant' });
          }
        }
      }

      // module.exports
      else if (line.match(/^\s*module\.exports\s*=\s*/)) {
        exports.push({ name: 'default', type: 'default' });
      }
      
      // exports.name
      else if (line.match(/^\s*exports\.(\w+)\s*=/)) {
        const match = line.match(/^\s*exports\.(\w+)\s*=/);
        if (match) {
          exports.push({ name: match[1], type: 'constant' });
        }
      }
    }

    return exports;
  }

  private parseConstants(): ConstantInfo[] {
    const constants: ConstantInfo[] = [];
    
    const pattern = /^\s*(?:export\s+)?(?:const|let|var)\s+(\w+)\s*=\s*(.+)/;

    for (const line of this.lines) {
      const match = line.match(pattern);
      if (match) {
        constants.push({
          name: match[1],
          value: match[2]?.trim()
        });
      }
    }

    return constants;
  }

  private parseParameters(paramString: string): string[] {
    if (!paramString.trim()) return [];
    
    return paramString.split(',').map(param => {
      // Remove default values for simplicity
      return param.split('=')[0].trim();
    }).filter(param => param.length > 0);
  }

  private findFunctionEnd(startLine: number): number {
    // Look for arrow functions
    if (this.lines[startLine].includes('=>')) {
      // Single line arrow function
      if (!this.lines[startLine].includes('{')) {
        return startLine + 1;
      }
    }
    
    return this.findEndLine(startLine + 1, '{', '}');
  }

  private parseClassMethods(startLine: number, endLine: number): FunctionInfo[] {
    const methods: FunctionInfo[] = [];
    
    for (let i = startLine - 1; i < Math.min(endLine, this.lines.length); i++) {
      const line = this.lines[i];
      
      // Method patterns
      const patterns = [
        /^\s*(?:async\s+)?(\w+)\s*\(([^)]*)\)\s*\{/,  // method() {}
        /^\s*(\w+)\s*:\s*(?:async\s+)?\(([^)]*)\)\s*=>/,  // method: () => {}
        /^\s*(\w+)\s*:\s*(?:async\s+)?function\s*\(([^)]*)\)\s*\{/  // method: function() {}
      ];
      
      for (const pattern of patterns) {
        const match = line.match(pattern);
        if (match) {
          const name = match[1];
          const params = this.convertToParameterInfo(this.parseParameters(match[2] || ''));
          const isAsync = line.includes('async');
          const lineStart = i + 1;
          const lineEnd = this.findFunctionEnd(i);
          const docstring = this.extractDocstring(Math.max(0, i - 2), i);

          methods.push({
            name,
            parameters: params,
            isAsync,
            lineStart,
            lineEnd,
            docstring
          });
          break;
        }
      }
    }

    return methods;
  }

  private parseClassProperties(startLine: number, endLine: number): string[] {
    const properties: string[] = [];
    
    for (let i = startLine - 1; i < Math.min(endLine, this.lines.length); i++) {
      const line = this.lines[i];
      
      // Property patterns - simple assignment in constructor or class body
      const patterns = [
        /^\s*this\.(\w+)\s*=/,  // this.property = value
        /^\s*(\w+)\s*=\s*[^(]/   // property = value (not a method)
      ];
      
      for (const pattern of patterns) {
        const match = line.match(pattern);
        if (match && !line.includes('(')) {
          properties.push(match[1]);
          break;
        }
      }
    }

    return properties;
  }


  private convertToParameterInfo(params: string[]): ParameterInfo[] {
    return params.map(param => ({
      name: param,
      optional: false,
      destructured: false
    }));
  }

  private convertToPropertyInfo(props: string[]): PropertyInfo[] {
    return props.map(prop => ({
      name: prop,
      visibility: 'public' as const
    }));
  }

  private convertToImportItems(items: string[]): ImportItem[] {
    return items.map(item => ({
      name: item
    }));
  }
}